### PR TITLE
Allow to assumeRole from ConfigurationProvider

### PR DESCRIPTION
**Pull request by amenophis**
This PR aim to allow use of assumeRole with the ConfigurationProvider.

- [X] POC
- [X] Documentation

### CODE REVIEW

File: src/Core/src/Credentials/ConfigurationProvider.php [hunk 1]
```diff
@@ -29,4 +51,37 @@ public function getCredentials(Configuration $configuration): ?Credentials
 
         return null;
     }
+
+    private function getCredentialsFromRole(string $roleArn, ?string $sessionName, ?string $region): ?Credentials
+    {
+        $sessionName = $sessionName ?? \uniqid('async-aws-', true);
+
+        $stsClient = new StsClient(['region' => $region], new NullProvider(), $this->httpClient);
```
Comment (jderusse) [inline src/Core/src/Credentials/ConfigurationProvider.php:59]: I think `NullProvider` wont work. You need a key/secret somewhere.

It should be something similar to what have been generated in `getCredentials` method
```php
new Credentials(
  $configuration->get(Configuration::OPTION_ACCESS_KEY_ID),
  $configuration->get(Configuration::OPTION_SECRET_ACCESS_KEY),
  $configuration->get(Configuration::OPTION_SESSION_TOKEN)
);
```

**Review by jderusse — COMMENTED** [review]
Wow.. Thank you for this PR.
I've just a small coment, otherwise everything looks perect to me.

Did you tried it? If no let me time to check it

**Comment by Nyholm** [comment]
Thank you. 

I would like to see some docs too.

**Comment by amenophis** [comment]
Yes, i would like to know if all is OK, i will add documentation after :)

**Comment by amenophis** [comment]
It seems the ConfigurationProvider is not documented. I miss something ?

File: src/Core/src/Credentials/ConfigurationProvider.php [hunk 2]
```diff
@@ -13,20 +17,71 @@
...
+        if (null === $accessKeyId || null === $secretAccessKeyId) {
+            return null;
+        }
+
+        $roleArn = $configuration->get(Configuration::OPTION_ROLE_ARN);
+        if (null !== $roleArn) {
+            $region = $configuration->get(Configuration::OPTION_REGION);
+            $roleSessionName = $configuration->get(Configuration::OPTION_ROLE_SESSION_NAME);
+
+            return $this->getCredentialsFromRole($accessKeyId, $secretAccessKeyId, $region, $roleArn, $roleSessionName);
```
Comment (jderusse) [inline src/Core/src/Credentials/ConfigurationProvider.php:44]: Shouldn't we provide the sessionToken (when defined) here?

```php
$credentials = new Credentials(
    $configuration->get(Configuration::OPTION_ACCESS_KEY_ID),
    $configuration->get(Configuration::OPTION_SECRET_ACCESS_KEY),
    $configuration->get(Configuration::OPTION_SESSION_TOKEN)
);
if (null === $roleArn) {
   return $credentials;
}

return $this->getCredentialsFromRole($credentials, ...);

```

Comment (amenophis) [reply src/Core/src/Credentials/ConfigurationProvider.php:44]: I can pass the session token to the getCredentialsFromRole method if defined, and use add it ton the Credentials for StsClient.
Is it your point ?

Comment (amenophis) [reply src/Core/src/Credentials/ConfigurationProvider.php]: Pushed the update, is it better ?

**Review by jderusse — APPROVED** [review]
Tried with personnal account, and work as expected.

Could you please just add a note (or sample) in documentation in `hard-coded.md`
The documentation in `configuration.md` should be fixed (not only related to WebIdentity Provider)

**Comment by amenophis** [comment]
Documentation pushed.

File: docs/authentication/hard-coded.md [hunk 3]
```diff
@@ -7,6 +7,7 @@ category: authentication
 When developing, and debugging, the simplest way to configure the client, is to set the credentials in the
 client configuration parameters.
 
+## Connect with accessKey / accessKeySecret
```
Comment (Nyholm) [inline docs/authentication/hard-coded.md:10]:
```suggestion
## Connect with access key and secret

```

File: docs/configuration.md [hunk 4]
```diff
@@ -85,8 +85,7 @@ for authentication.
 
 ### roleArn
 
-The Amazon Resource Name (ARN) of the role that the caller is assuming when using
-the [WebIdentity Provider](/authentication/web-identity.md)
+The Amazon Resource Name (ARN) of the role that the caller is assuming.
```
Comment (Nyholm) [inline docs/configuration.md:88]:
```suggestion
The Amazon Resource Name (ARN) of the role that the client should be "assuming" after authentication. 
```

File: src/Core/src/Credentials/ConfigurationProvider.php [hunk 5]
```diff
@@ -13,20 +17,72 @@
  */
 final class ConfigurationProvider implements CredentialProvider
 {
+    private $logger;
+
+    private $httpClient;
+
+    public function __construct(?LoggerInterface $logger = null, ?HttpClientInterface $httpClient = null)
```
Comment (Nyholm) [inline src/Core/src/Credentials/ConfigurationProvider.php:24]: Let's flip these arguments to be constant with other classes. 

```suggestion
    public function __construct(?HttpClientInterface $httpClient = null, ?LoggerInterface $logger = null)
```

File: docs/authentication/hard-coded.md [hunk 6]
```diff
@@ -16,5 +17,19 @@ $client = new AwsClientFactory([
 ]);
 ```
 
+## Connect with accessKey / accessKeySecret + assumeRole
+You can use a specific role (assumeRole) by specify a roleArn.
+You can also specify a roleSessionName.
```
Comment (Nyholm) [inline docs/authentication/hard-coded.md:22]:
```suggestion
## Assume role ARN with access key and secret

You can tell the API client to use specific role by providing a ARN to 
the `roleArn` key. You can also specify a `roleSessionName`.
```

**Review by Nyholm — COMMENTED** [review]
I think this PR looks good. 

I added some suggestions. Review then and change them if you like.

**Comment by amenophis** [comment]
All good !

**Review by Nyholm — APPROVED** [review]
Im happy with this PR. I'll let @jderusse do a final review and merge if he likes it.

File: docs/authentication/hard-coded.md [hunk 7]
```diff
@@ -7,6 +7,8 @@ category: authentication
 When developing, and debugging, the simplest way to configure the client, is to set the credentials in the
 client configuration parameters.
 
+## Connect with access key and secret
```
Comment (jderusse) [inline docs/authentication/hard-coded.md:10]:
```suggestion
## Authenticate with access key and secret
```

File: docs/authentication/hard-coded.md [hunk 8]
```diff
@@ -16,5 +18,20 @@ $client = new AwsClientFactory([
 ]);
 ```
 
+## Assume role ARN with access key and secret
```
Comment (jderusse) [inline docs/authentication/hard-coded.md:21]:
```suggestion
## Assuming a role
```

**Review by jderusse — APPROVED** [review]
Thanks for this contribution

**Pull request merged**